smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/CodegenVisitor.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.go.codegen;
+
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.logging.Logger;
+import software.amazon.smithy.build.FileManifest;
+import software.amazon.smithy.build.PluginContext;
+import software.amazon.smithy.codegen.core.SymbolProvider;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.neighbor.Walker;
+import software.amazon.smithy.model.shapes.ServiceShape;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.shapes.ShapeVisitor;
+import software.amazon.smithy.model.shapes.StructureShape;
+
+/**
+ * Orchestrates Go client generation.
+ */
+final class CodegenVisitor extends ShapeVisitor.Default<Void> {
+
+    private static final Logger LOGGER = Logger.getLogger(CodegenVisitor.class.getName());
+
+    private final GoSettings settings;
+    private final Model model;
+    private final Model modelWithoutTraitShapes;
+    private final ServiceShape service;
+    private final FileManifest fileManifest;
+    private final SymbolProvider symbolProvider;
+    private final GoDelegator writers;
+
+    CodegenVisitor(PluginContext context) {
+        settings = GoSettings.from(context.getModel(), context.getSettings());
+        model = context.getModel();
+        modelWithoutTraitShapes = context.getModelWithoutTraitShapes();
+        service = settings.getService(model);
+        fileManifest = context.getFileManifest();
+        LOGGER.info(() -> "Generating Go client for service " + service.getId());
+
+        symbolProvider = GoCodegenPlugin.createSymbolProvider(model);
+        writers = new GoDelegator(settings, model, fileManifest, symbolProvider);
+    }
+
+    void execute() {
+        // Generate models that are connected to the service being generated.
+        LOGGER.fine("Walking shapes from " + service.getId() + " to find shapes to generate");
+        Set<Shape> serviceShapes = new TreeSet<>(new Walker(modelWithoutTraitShapes).walkShapes(service));
+
+        for (Shape shape : serviceShapes) {
+            shape.accept(this);
+        }
+
+        LOGGER.fine("Flushing go writers");
+        writers.flushWriters();
+
+        LOGGER.fine("Generating go.mod file");
+        GoModGenerator.writeGoMod(settings, fileManifest);
+    }
```

Comment:
I'm a bit on the fence still on whether we should be generating to the go module file as part of the smithy type generation. Especially since at the time of generating this go mod file the complete dependency enclosure is not known from a type perspective, additionally it's not clear to me whether we should be treating this file as a code-generated artifact that is rewritten every generation.

Reply:
> at the time of generating this go mod file the complete dependency enclosure is not known

Sure it is, why wouldn't it be? Code generator plugins will have to register their dependencies.

> it's not clear to me whether we should be treating this file as a code-generated artifact that is rewritten every generation

It's what defines a module, why wouldn't we generate it every time? The code generator has no notion of existing state, it outputs to an ephemeral directory that could be wiped at any time that doesn't have version control. From the SDK publishing perspective you would then copy that out into its final home.

Reply:
We talked about this offline. We're still going to generate it, but we're going to use the Go command line tooling to do so.